File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-3.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        if len(nums) == 1: return 0 if nums[0] == target else -1

        l, r = 0, len(nums)-1
        
        while l < r-1:
            m = l + ((r-l)//2)
            if nums[m] > nums[l]:
                l = m
            elif nums[m] < nums[r]:
                r = m

        indexDiff = len(nums)-r if nums[0] > nums[-1] else 0

        l, r = 0, len(nums)-1

        while l <= r:
            m = l + ((r-l)//2)
            if nums[m-indexDiff] < target:
                l = m+1
            elif nums[m-indexDiff] > target:
                r = m-1
            else:
                return (m-indexDiff) if (m-indexDiff) >= 0 else m + (len(nums)-indexDiff)
        return -1

File: Data Structures & Algorithms/4sum/submission-3.cpp
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());

        kSum(nums, 4, 0, (long long) target);

        return ans;
    }

private:
    vector<vector<int>> ans;
    vector<int> quad;

    vector<vector<int>> kSum(vector<int>& nums, int k, int i, long long target) {
        int n = nums.size();

        if (k == 2) {

            int l = i, r = n - 1;

            while (l < r) {
                long long sum = (long long) nums[l] + nums[r];
                if (sum == target) {
                    quad.push_back(nums[l]);
                    quad.push_back(nums[r]);
                    ans.push_back(quad);
                    quad.pop_back();
                    quad.pop_back();
                    do {
                        l++;
                    } while (l < r && nums[l] == nums[l - 1]);
                    do {
                        r--;
                    } while (l < r && nums[r] == nums[r + 1]);
                } else if (sum < target) {
                    l++;
                } else {
                    r--;
                }
            }

            return ans;

        }

        for (int j = i; j < n; j++) {
            if (j > i && nums[j] == nums[j - 1]) continue;

            quad.push_back(nums[j]);
            kSum(nums, k - 1, j + 1, target - nums[j]);
            quad.pop_back();
        }

        return ans;

    }
};